core/src/main/java/org/apache/iceberg/PositionDeletesTable.java

```diff
@@ -0,0 +1,221 @@
+                MetadataColumns.FILE_PATH_COLUMN_ID,
+                "delete_file_path",
+                Types.StringType.get(),
+                MetadataColumns.FILE_PATH_COLUMN_DOC));
+
+    if (partitionType.fields().size() > 0) {
+      return result;
+    } else {
+      // avoid returning an empty struct, which is not always supported.
+      // instead, drop the partition field
+      return TypeUtil.selectNot(result, Sets.newHashSet(MetadataColumns.PARTITION_COLUMN_ID));
+    }
+  }
+
+  public static class PositionDeletesBatchScan
+      extends SnapshotScan<BatchScan, ScanTask, ScanTaskGroup<ScanTask>> implements BatchScan {
+
+    protected PositionDeletesBatchScan(Table table, Schema schema) {
+      super(table, schema, new TableScanContext());
+    }
+
+    protected PositionDeletesBatchScan(Table table, Schema schema, TableScanContext context) {
+      super(table, schema, context);
+    }
+
+    @Override
+    protected PositionDeletesBatchScan newRefinedScan(
+        Table newTable, Schema newSchema, TableScanContext newContext) {
+      return new PositionDeletesBatchScan(newTable, newSchema, newContext);
+    }
+
+    @Override
+    public CloseableIterable<ScanTaskGroup<ScanTask>> planTasks() {
+      return TableScanUtil.planTaskGroups(
+          planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+    }
+
+    @Override
+    protected List<String> scanColumns() {
+      return context().returnColumnStats() ? DELETE_SCAN_WITH_STATS_COLUMNS : DELETE_SCAN_COLUMNS;
+    }
+
+    @Override
+    protected CloseableIterable<ScanTask> doPlanFiles() {
+      String schemaString = SchemaParser.toJson(tableSchema());
+
+      // prepare transformed partition specs and caches
+      Map<Integer, PartitionSpec> transformedSpecs =
+          table().specs().values().stream()
+              .map(spec -> transformSpec(tableSchema(), spec))
+              .collect(Collectors.toMap(PartitionSpec::specId, spec -> spec));
+
+      LoadingCache<Integer, ResidualEvaluator> residualCache =
+          partitionCacheOf(
+              transformedSpecs,
+              spec ->
+                  ResidualEvaluator.of(
+                      spec,
+                      shouldIgnoreResiduals() ? Expressions.alwaysTrue() : filter(),
+                      isCaseSensitive()));
+
+      LoadingCache<Integer, String> specStringCache =
+          partitionCacheOf(transformedSpecs, PartitionSpecParser::toJson);
+
+      LoadingCache<Integer, ManifestEvaluator> evalCache =
+          partitionCacheOf(
+              transformedSpecs,
+              spec -> ManifestEvaluator.forRowFilter(filter(), spec, isCaseSensitive()));
+
+      // iterate through delete manifests
+      CloseableIterable<ManifestFile> deleteManifests =
+          CloseableIterable.withNoopClose(snapshot().deleteManifests(table().io()));
+      CloseableIterable<ManifestFile> filteredManifests =
+          CloseableIterable.filter(
```

Comment:
Do we need to use a version of filter that updates `scanMetrics()`?

Reply:
Good catch, updated. Will add a test as we discussed.